### DISCUSSION TITLE
Add value names to enums

### DISCUSSION THREAD

**Discussion by JonathanStarup** [flix/flix#5329] [Ideas]
# Idea
I propose that you can write (This could be optional or mandatory)
```scala
enum Board {
    case Chess
    case Go(beginner: Bool)
    case Ludo(players: Int32)
    case FlammeRouge(seq: List[FlammeRougePiece])
    case Connect4(width: Int32, height: Int32)
```

### Documentation
Especially for terms of the same type you want to explain them. You cannot always put this confusion away into another type (what do you do with `Point(Int32, Int32)`).
#### Use cases
* [Channel.flix](https://github.com/flix/flix/blob/96c9bf9566f7d862d6e2640668959898f6b7913b/main/src/library/Concurrent/Channel.flix#L78) (names in comments)
* [MutDeque.flix](https://github.com/flix/flix/blob/96c9bf9566f7d862d6e2640668959898f6b7913b/main/src/library/MutDeque.flix#L18) (names in comments)
* [MutList.flix](https://github.com/flix/flix/blob/96c9bf9566f7d862d6e2640668959898f6b7913b/main/src/library/MutList.flix#L17) (ambiguity: Is the int ref capacity or length?)
* [Denotation.flix](https://github.com/flix/flix/blob/96c9bf9566f7d862d6e2640668959898f6b7913b/main/src/library/Fixpoint/Ast/Denotation.flix#L18) (names in comments)
* [PrecedenceGraph.flix](https://github.com/flix/flix/blob/96c9bf9566f7d862d6e2640668959898f6b7913b/main/src/library/Fixpoint/Ast/PrecedenceGraph.flix#L24) (ambiguity: what is from, what is to?)

#### Pattern Matching
Currently, a pattern match case of Connect4 autocompletes as `case Connect4(, ) => ???` and without names in the definition; how can you do better? The names would allow `case Connect4(width, height) => ???`, which is more helpful. 
#### Accessor template
With value names you could also generate named accessors for the terms automatically (`getXYZ`)

#### Direct Access
Named fields allow introducing term lookup like `Ludo(2).players` (just search `let [A-Z]\w+` in std lib and see the single term extractions)

**Comment by JonathanStarup** [comment]
I also remember multiple people asking about this in the public gitter channel, so it also removes friction from new users

**Comment by mlutze** [comment]
Not exactly the same, but I think we currently allow the "singleton record" pattern in enums.

**Comment by JonathanStarup** [comment]
Didn't know that, that's cool.
```scala
pub enum Test {
    case TestCase({x = Int32}, {y = Int32})
}

pub def example(): String = {
    stringify(TestCase(x = 32, y = 15))
}

pub def example2(t: Test): Int32 = {
    let TestCase(xr, yr) = t;
    xr.x + yr.y
}
```
It does have a lot of drawbacks though (typeclass derivation, pattern matching, and named arguments (if you don't want that)).

**Comment by magnus-madsen** [comment]
If we added type class derivation for records and pattern matching, do you think that would resolve most of your issues/concerns?
